src/Sparrow.Server/Allocators.Native.cs

```diff
@@ -0,0 +1,209 @@
+            //       JIT will be able to fold all this if sequence into a branchless single call.
+            if (allocator._options.ElectricFenceEnabled)
+                memory = ElectricFencedMemory.Allocate(size);
+            else if (allocator._options.UseSecureMemory)
+                throw new NotImplementedException();
+            else
+                memory = NativeMemory.AllocateMemory(size);
+
+            if (allocator._options.Zeroed)
+                Memory.Set(memory, 0, size);
+
+            Interlocked.Add(ref allocator._allocated, size);
+            Interlocked.Add(ref allocator._totalAllocated, size);
+
+            return new Pointer(memory, size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Release(ref NativeAllocator<TOptions> allocator, ref Pointer ptr)
+        {
+            Interlocked.Add(ref allocator._allocated, -ptr.Size);
+
+            // PERF: Given that for the normal use case the INativeOptions we will use returns constants the
+            //       JIT will be able to fold all this if sequence into a branchless single call.
+            if (allocator._options.ElectricFenceEnabled)
+                ElectricFencedMemory.Free((byte*)ptr.Address);
+            else if (allocator._options.UseSecureMemory)
+                throw new NotImplementedException();
+            else
+                NativeMemory.Free((byte*)ptr.Address, ptr.Size);
+
+            ptr = new Pointer();
+        }
+
+        public void Reset(ref NativeAllocator<TOptions> allocator)
+        {
+            // There is no reset action to do on this allocator. We wont fail, but nothing we can do to 'reclaim' or 'reuse' memory.            
+            CheckForLeaks(ref allocator);
+        }
+
+        public void OnAllocate(ref NativeAllocator<TOptions> allocator, Pointer ptr)
+        {
+            // This allocator does not keep track of anything.
+        }
+
+        public void OnRelease(ref NativeAllocator<TOptions> allocator, Pointer ptr)
+        {
+            // This allocator does not keep track of anything.
+        }
+
+        public void Dispose(ref NativeAllocator<TOptions> allocator, bool disposing)
+        {
+            // If we use the leak detector when finalizing an object, the dotnet process will explode
+            // with an unhandled exception. Therefore, we cannot check for leaks and throw in those cases. 
+            if (disposing)
+                CheckForLeaks(ref allocator);
+        }
+
+        [Conditional("DEBUG")]
+        [Conditional("VALIDATE")]
+        private void CheckForLeaks(ref NativeAllocator<TOptions> allocator)
+        {
+            if (allocator.Allocated != 0)
+            {
+#if VALIDATE || DEBUG
+
+                StringBuilder stackFrameAppender = new StringBuilder();
+                using (var reader = new StringReader(allocator._initializeStackTrace.ToString()))
```

Comment:
Why do you need this? Isn't this skipping the first three lines? if so, `Split().Skip(3)` is more readable and this is debug code

Reply:
Those 3 lines are the actual leak checker code, skipping them is more readable. From the code point of view, yeah Split().Skip(3) would do the trick.